staging/src/k8s.io/kms/encryption/managed.go

```diff
@@ -0,0 +1,264 @@
+package encryption
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+const (
+	// MaxUsage with 2^21 is a very defensive value. 2^32 is more commonly used.
+	MaxUsage = 2097151
+	// keySize is the key size in bytes
+	keySize = 128 / 8
+	// nonceSize is the size of the nonce. Do not change, without breaking version change.
+	// The nonceSize is a de facto standard.
+	nonceSize = 12
+	// cacheSize is set to 100 keys, which can be used to decrypt up to 200 million encrypted
+	// files or data encrypted within nearly 2 years.
+	cacheSize = 100
+)
+
+var (
+	// ErrKeyExpired means that the expiration time of a key has come and it shouldn't be used any more.
+	ErrKeyExpired = errors.New("key is out of date and shouldn't be used anymore for encryption")
+	// ErrNoCipher means that there is no remote kms given and therefore the keys in use can't be protected.
+	ErrNoCipher = errors.New("no remote encryption service was specified")
+
+	week = time.Hour * 24 * 7
+)
+
+// ManagedCipher is a set of keys. Only one key is used for encryption at one
+// time. New keys are created automatically, when hitting safety thresholds.
+type ManagedCipher struct {
+	counter uint32
+	expires time.Time
+
+	keys *cache
+
+	currentRemoteKMSID []byte
+	currentLocalKEK    []byte
+
+	fallbackRemoteKMSID []byte
+	fallbackLocalKEK    []byte
+
+	remoteKMS EncrypterDecrypter
+
+	m sync.Mutex
+}
+
+// EncrypterDecrypter is a default encryption / decryption interface with an ID
+// to support remote state.
+type EncrypterDecrypter interface {
+	// TODO: consider adding context to interface.
+	Encrypt(plaintext []byte) (keyID, ciphertext []byte, err error)
+	Decrypt(keyID, ciphertext []byte) (plaintext []byte, err error)
+}
+
+// CurrentKeyID returns the currently assumed remote Key ID.
+func (m *ManagedCipher) CurrentKeyID() []byte {
+	return m.currentRemoteKMSID
+}
+
+// NewManagedCipher returns a pointer to a ManagedCipher. It is initialized with
+// a cache, a reference to an remote cipher (like a KMS service or HMS) and
+// does an initial encryption call to the remote cipher.
+func NewManagedCipher(remoteCipher EncrypterDecrypter) (*ManagedCipher, error) {
+	if remoteCipher == nil {
+		klog.Infof("create managed cipher without remote encryption failed")
```

Comment:
Should we log these at higher log levels instead of by default? (relevant for all logs in this file)

Reply:
So we set it to debugging level `V(4)`, as it might be used for debugging or should it be a `klog.Errorf`?